ts/private/ts_project_worker.js

```diff
@@ -29,85 +29,138 @@ function getArgsFromParamFile() {
 function createFilesystemTree(root, inputs) {
     const hashes = new Map();
     const tree = {};
+    const watchingTree = {};
+
+    const TYPE = {
+        DIR: 1,
+        FILE: 2,
+        SYMLINK: 3
+    }
+
+    const EVENT_TYPE = {
+        ADDED: 0,
+        UPDATED: 1,
+        REMOVED: 2,
+    }
+
+    const Type = Symbol.for("fileSystemTree#type");
+    const Symlink = Symbol.for("fileSystemTree#symlink");
+    const Watcher = Symbol.for("fileSystemTree#watcher");
 
     for (const p in inputs) {
         add(p, inputs[p]);
     }
 
     function getNode(p) {
         const parts = p.split(path.sep);
-        let backtrack = tree;
-    
+        let node = tree;
         for (const part of parts) {
-            backtrack = backtrack[part];
-            if (!backtrack) {
+            if (!(part in node)) {
                 return undefined;
             }
-            if (backtrack.symlinkTo) {
-                backtrack = getNode(backtrack.symlinkTo);
+            node = node[part];
+            if (node[Type] == TYPE.SYMLINK) {
+                node = getNode(node[Symlink]);
             }
         }
-        return backtrack;
+        return node;
     }
 
     function add(p, hash) {
-        const parts = p.split(path.sep);
-        let backtrack = tree;
-        for (const part of parts) {
-            if (typeof backtrack[part] != "object") {
-                backtrack[part] = {};
+        const parts = path.parse(p);        
+        const dirs = parts.dir.split(path.sep).filter(p => p != "");
```

Comment:
Do we actually need the `.filter`? I'd think if the path is normalized it wouldn't be necessary.

Reply:
problem is that `.split` will return the whole string when there is no match for the `sep`. 

for instance
```
const path = require("path");
const p = path.parse("dir");
const dirs = p.dir.split(path.sep);
console.log(p, dirs);
```

`p.dir.split(path.sep)` will yield `[""]`

Reply:
https://codesandbox.io/s/node-playground-forked-9dyrpc

Reply:
Ah I see, I didn't notice what the `parts.dir` was and assumed it was `p.split("/")`...